fixieai/agents/agent_func.py

```diff
@@ -232,26 +307,93 @@ def create(
         )
 
 
-def _adapt_func_result(
-    result: Any, allow_multiple: bool
-) -> Iterable[api.AgentResponse]:
-    """Adapts any allowed func return type into an Iterable of AgentResponses."""
-
-    if isinstance(result, api.AgentResponse):
-        yield result
-    elif isinstance(result, api.Message):
-        yield api.AgentResponse(message=result)
-    elif isinstance(result, str):
-        yield api.AgentResponse(message=api.Message(result))
-    elif isinstance(result, collections.abc.Iterable):
-        for value in result:
-            yield from _adapt_func_result(value, allow_multiple)
-
-            if not allow_multiple:
-                break
-    else:
-        raise TypeError(
-            f"The func result was type {type(result)}, but must be a str, Message, AgentResponse, or iterable."
+class AgentCorpusFunc(
+    AgentFunc[
+        corpora.CorpusRequest,
+        corpora.CorpusResponse,
+        corpora.CorpusResponse,
+    ]
+):
+    """A Python function that can be invoked by Fixie to load a custom corpus.
+
+    Wrapped functions must take one `corpora.CorpusRequest` argument (or no
+    arguments) and return a `corpora.CorpusResponse`.
+    """
+
+    def adapt_result(
+        self, result: CorpusResponse
+    ) -> Generator[CorpusResponse, None, None]:
+        if isinstance(result, CorpusResponse):
+            yield result
+        else:
+            raise TypeError(
+                f"The func result was type {type(result)}, but must be a CorpusResponse."
+            )
+
+    def adapt_exception(self, e: exceptions.AgentException) -> CorpusResponse:
+        logging.warn(f"Corpus loading threw an exception: {e}", stack_info=True)
+        return corpora.CorpusResponse()
```

Comment:
I think this should probably just raise and HTTP 500 -- unlike `AgentQuery`/`AgentReponse` there's not a human or LLM on the other end of the request.